sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetUtils.scala

```diff
@@ -127,4 +145,292 @@ object ParquetUtils {
+  private[sql] def getPushedDownAggResult(
+      footer: ParquetMetadata,
+      dataSchema: StructType,
+      aggregation: Aggregation)
+  : (Array[PrimitiveType.PrimitiveTypeName], Array[Any]) = {
+    val footerFileMetaData = footer.getFileMetaData
+    val fields = footerFileMetaData.getSchema.getFields
+    val blocks = footer.getBlocks()
+    val typesBuilder = ArrayBuilder.make[PrimitiveType.PrimitiveTypeName]
+    val valuesBuilder = ArrayBuilder.make[Any]
+
+    for (i <- 0 until aggregation.aggregateExpressions.size) {
+      var value: Any = None
+      var rowCount = 0L
+      var isCount = false
+      var index = 0
+      blocks.forEach { block =>
+        val blockMetaData = block.getColumns()
+        aggregation.aggregateExpressions(i) match {
+          case Seq(Max(col, _)) =>
+            index = dataSchema.fieldNames.toList.indexOf(col)
+            val currentMax = getCurrentBlockMaxOrMin(footer, blockMetaData, index, true)
+            if (currentMax != None &&
+              (value == None || currentMax.asInstanceOf[Comparable[Any]].compareTo(value) > 0)) {
+              value = currentMax
+            }
+
+          case Seq(Min(col, _)) =>
+            index = dataSchema.fieldNames.toList.indexOf(col)
+            val currentMin = getCurrentBlockMaxOrMin(footer, blockMetaData, index, false)
+            if (currentMin != None &&
+              (value == None || currentMin.asInstanceOf[Comparable[Any]].compareTo(value) < 0)) {
+              value = currentMin
+            }
+
+          case Seq(Count(col, _, _)) =>
+            index = dataSchema.fieldNames.toList.indexOf(col)
+            rowCount += block.getRowCount
+            if (!col.equals("1")) {  // "1" is for count(*)
+              rowCount -= getNumNulls(footer, blockMetaData, index)
+            }
+            isCount = true
+
+          case _ =>
+        }
+      }
+      if (isCount) {
+        valuesBuilder += rowCount
+        typesBuilder += PrimitiveType.PrimitiveTypeName.INT96
+      } else {
+        valuesBuilder += value
+        typesBuilder += fields.get(index).asPrimitiveType.getPrimitiveTypeName
+      }
+    }
+    (typesBuilder.result(), valuesBuilder.result())
+  }
+
+  /**
+   * get the Max or Min value for ith column in the current block
+   *
+   * @return the Max or Min value
+   */
+  private def getCurrentBlockMaxOrMin(
+      footer: ParquetMetadata,
+      columnChunkMetaData: util.List[ColumnChunkMetaData],
+      i: Int,
+      isMax: Boolean): Any = {
+    val parquetType = footer.getFileMetaData.getSchema.getType(i)
+    if (!parquetType.isPrimitive) {
+      throw new IllegalArgumentException("Unsupported type : " + parquetType.toString)
+    }
+    val statistics = columnChunkMetaData.get(i).getStatistics()
+    if (!statistics.hasNonNullValue) {
+      throw new UnsupportedOperationException("No min/max found for parquet file, Set SQLConf" +
+        " PARQUET_AGGREGATE_PUSHDOWN_ENABLED to false and execute again")
```

Comment:
We cannot avoid this exception on runtime? It looks inconvenient. For example, in `SupportsPushDownAggregates .pushAggregation`, we cannot give up pushing down aggregates in case of this unsupported one?

Reply:
This part bothers me too. I don't know why in some cases parquet doesn't return statistics info. We won't be able to find out until we read footer in `FilePartitionReaderFactory.createReader`. It's already passed the `SupportsPushDownAggregates .pushAggregation`. I can't figure out a good way to do this.